### PR TITLE
[C] do not reapply a Binding with a Source when the Context changes

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
### Description of Change ###

[C] do not reapply a Binding with a Source:
- when the Context changes,
- the view is parented
- the inherited context changes.

### Bugs Fixed ###

- https://bugzilla.xamarin.com/show_bug.cgi?id=57081

### API Changes ###

/

### Behavioral Changes ###

/

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

**Comment by StephaneDelcroix** [comment]
@jassmith please review the alternate fix. the original one was blocking reapplying from all changes (propertychanged, ...)

**Comment by StephaneDelcroix** [comment]
> @jimmgarrido: @stephane @davidortinau the updated #983 does not resolve https://bugzilla.xamarin.com/show_bug.cgi?id=57081. The original fix that had failing tests did resolve it however.

**Comment by Szymaniuk** [comment]
In which Xamarin Forms release this COULD be released?
Thanks.

**Comment by StephaneDelcroix** [comment]
@jimmgarrido was right, this doesn't fix the problem entirely

File: Xamarin.Forms.Core/Binding.cs [hunk 1]
```diff
@@ -114,10 +114,15 @@ internal override void Apply(bool fromTarget)
...
-		internal override void Apply(object newContext, BindableObject bindObj, BindableProperty targetProperty)
+		internal override void Apply(object newContext, BindableObject bindObj, BindableProperty targetProperty, bool fromBindingContextChanged = false)
 		{
 			object src = _source;
-			base.Apply(src ?? newContext, bindObj, targetProperty);
+			var isApplied = IsApplied;
+
+			base.Apply(src ?? newContext, bindObj, targetProperty, fromBindingContextChanged: fromBindingContextChanged);
+
+			if (src != null && isApplied && fromBindingContextChanged)
```
Comment (kingces95) [inline Xamarin.Forms.Core/Binding.cs]: If I remove `fromBindingContextChanged` and change this line to simply `if (src != null && isApplied)` then all 80 unit tests in `BindableObjectUnitTests.cs`, including the two new ones, all pass. Let's include a test that fails.

**Comment by StephaneDelcroix** [comment]
@kingces95 I can't write a test that fails without the `&& fromBindingContextChanged`, so the check might not be required. I still choose to leave it in as it shows the intent.

**Comment by mariusbloemhof** [comment]
Is anyone aware of the fact that this issue has resurfaced from version 3.1?

It was working in version 2.5.1.52 and 3.0. I upgraded to 3.1 and we are seeing duplicate view model creation using navigation service

**Comment by StephaneDelcroix** [comment]
@mariusbloemhof the unit test we added with this PR are still passing. Could you please open a new issue with a simple working reproduction so we can investigate ?

Thanks a lot

**Pull request merged**